llvm/lib/Target/AMDGPU/AMDGPUCloneModuleLDS.cpp

```diff
@@ -0,0 +1,139 @@
+//===-- AMDGPUCloneModuleLDSPass.cpp ------------------------------*- C++ -*-=//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// The purpose of this pass is to ensure that the combined module contains
```

Comment:
What the pass does and why it's needed is still unclear to me, so I have several concerns:

If I understand correctly, this is only needed for the LTO pipeline so I'd add a comment to say this runs before LowerModuleLDS in the FullLTO pipeline, and is used to improve eventual usage of --lto-partitions.

Does the pass affect codegen in any way when --lto-partitions is not used? If you add it everytime before LowerModuleLDS, does it change codegen in any tests?

I also don't see how this improves splitting. I guess it makes more isolated partitions if we clone functions, but #89245 already clones functions that are small enough and it has heuristics built in to minimize duplication for larger kernels/dependencies tree.

Reply:
I updated the description, I hope the purpose is clear now.

The cloning utilities that you are using in your patch don't clone LDS globals. Instead, they are externalized in the new modules. This only results in too many unnecessary dependencies across modules which will in turn lower the target performance of parallel compilation.